model.lua

```diff
@@ -150,33 +150,57 @@ function Model:retrieve(dataloader, dtype)
 
     self.params.numOptions = 100;
 
-    -- ranks for the given datatype
-    local ranks = torch.Tensor(numThreads, self.params.maxQuesCount);
-    ranks:fill(self.params.numOptions + 1);
-
-    while startId <= numThreads do
-        -- print progress
-        xlua.progress(startId, numThreads);
-
-        -- grab a batch
-        local batch, nextStartId =
-                        dataloader:getTestBatch(startId, self.params, dtype);
-
-        -- Call retrieve function for specific model, and store ranks
-        ranks[{{startId, nextStartId - 1}, {}}] = self:retrieveBatch(batch);
-        startId = nextStartId;
-    end
+    local retrievalOutput;
+    if dtype == 'val' then
+        -- ranks for the val split, test split has no ground truth to compare ranks
+        local ranks = torch.Tensor(numThreads, self.params.maxQuesCount);
+        ranks:fill(self.params.numOptions + 1);
+
+        while startId <= numThreads do
+            -- print progress
+            xlua.progress(startId, numThreads);
+
+            -- grab a batch
+            local batch, nextStartId =
+                            dataloader:getTestBatch(startId, self.params, dtype);
+
+            -- Call retrieve function for specific model, and store ranks
+            ranks[{{startId, nextStartId - 1}, {}}] = self:retrieveBatch(batch, dtype);
+            startId = nextStartId;
+        end
 
-    print(string.format('\n%s - Retrieval:', dtype))
-    utils.processRanks(ranks);
+        print(string.format('\n%s - Retrieval:', dtype))
+        utils.processRanks(ranks);
+        retrievalOutput = ranks;
+    else
+        -- sorted answer options of last round for test split
+        local answerOptions = torch.zeros(numThreads, self.params.numOptions);
+
+        while startId <= numThreads do
+            -- print progress
+            xlua.progress(startId, numThreads);
+
+            -- grab a batch
+            local batch, nextStartId =
+                            dataloader:getTestBatch(startId, self.params, dtype);
+
+            -- Call retrieve function for specific model, and store ranks
+            local batchAnswerOptions = self:retrieveBatch(batch, dtype);
+            -- remove dummy answer options for first nine rounds
+            batchAnswerOptions = batchAnswerOptions:view(nextStartId - startId, -1, self.params.numOptions)
+            answerOptions[{{startId, nextStartId - 1}, {}}] = batchAnswerOptions[{{}, {10}}];
```

Comment:
Unless I am wrong, shouldn't you be picking the round based on the round for the current test instance? I remember we padding zeros to later part of the dialog. For instance, if the test instance is at round 5, we pad rounds 6..10 as 0, and need to pick the rank at round 5, correct?

Reply:
@satwikkottur I guess you reviewed on an outdated diff. Please review on the latest commit - there are two methods `retrieve` and `predict`. 
While this line stays, the ranks list of last round is picked out while forming the retrieval table (which is ready to be saved as JSON).
This line in particular saves the rank list at `round_id` index of ranks table: https://github.com/batra-mlp-lab/visdial/pull/18/commits/83cfa1f86d1aeb5a3bd8f9446d7014600535637e#diff-cd27003c24a2c135a29b100aa893b075R236